File: apps/workers/internal/schedule/worker.go
package schedule

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"time"

	enumspb "go.temporal.io/api/enums/v1"
	"go.temporal.io/sdk/client"
	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/worker"
)

// Options is what the worker needs, supplied by main.
type Options struct {
	// Addr is the engine's frontend, host:port.
	Addr string
	// Namespace the schedules live in.
	Namespace string
	// TaskQueue this worker polls; `core` on the bundled stack.
	TaskQueue string
	// SnoozeExpirySchedule is a five-field cron expression, evaluated in UTC.
	SnoozeExpirySchedule string
	// OutboxRelayInterval is how often the relay looks for pending mail.
	OutboxRelayInterval time.Duration
	// OutboxReclaimSchedule is a five-field cron expression, evaluated in UTC.
	OutboxReclaimSchedule string
	// SweepRelayInterval is how often the relay looks for sweeps somebody
	// asked for.
	SweepRelayInterval time.Duration
	// SweepSchedule is a five-field cron expression, evaluated in UTC, for
	// sweeping every organisation.
	SweepSchedule string
	// ExecutorRelayInterval is how often the relay looks for approvals whose
	// record has not been created yet.
	ExecutorRelayInterval time.Duration
	// Activities is the dependency set the activities close over.
	Activities *Activities
	Logger     *slog.Logger
}

// SnoozeExpiryScheduleID is the Schedule's id in the engine, which is what an
// operator sees in `temporal schedule list` and the UI, and what the CI step
// asks for.
const SnoozeExpiryScheduleID = "expire-snoozed-findings"

// Worker is a connected client plus a running worker, and the handle main uses
// to stop it.
type Worker struct {
	client client.Client
	worker worker.Worker
}

// Connect dials the engine, retrying until it answers or the context ends.
//
// Retried for the same reason core-api retries OIDC discovery: `temporal` and
// `workers` start together, and losing that race is ordinary rather than
// exceptional. The compose file orders them, but a compose file is one
// deployment and the binary should not depend on it.
func Connect(ctx context.Context, opts Options) (client.Client, error) {
	const attempts = 30
	var lastErr error
	for attempt := 1; attempt <= attempts; attempt++ {
		c, err := client.DialContext(ctx, client.Options{
			HostPort:  opts.Addr,
			Namespace: opts.Namespace,
			Logger:    slogAdapter{opts.Logger},
		})
		if err == nil {
			return c, nil
		}
		lastErr = err
		opts.Logger.Info("waiting for temporal", "addr", opts.Addr, "attempt", attempt, "error", err)
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-time.After(2 * time.Second):
		}
	}
	return nil, fmt.Errorf("schedule: temporal at %s did not answer after %d attempts: %w", opts.Addr, attempts, lastErr)
}

// Start registers the workflows and activities, starts polling the task
// queue, and makes sure the schedules exist.
//
// Schedules are created here rather than by an operator, because a deployment
// where the engine is up and nobody ran the command is a deployment where
// nothing is scheduled, which looks exactly like a working one. The worker
// that serves a schedule is the thing that knows it should exist.
func Start(ctx context.Context, c client.Client, opts Options) (*Worker, error) {
	// The relay starts delivery workflows through the client it is given, on
	// the queue this worker polls. Set here rather than by main, so the two
	// cannot name different queues.
	opts.Activities.Starter = c
	opts.Activities.TaskQueue = opts.TaskQueue

	w := worker.New(c, opts.TaskQueue, worker.Options{})
	w.RegisterWorkflow(ExpireSnoozedFindingsWorkflow)
	w.RegisterWorkflow(RelayOutboxWorkflow)
	w.RegisterWorkflow(DeliverMessageWorkflow)
	w.RegisterWorkflow(ReclaimOutboxWorkflow)
	w.RegisterWorkflow(DeliverNotificationWorkflow)
	w.RegisterWorkflow(RelaySweepTriggersWorkflow)
	w.RegisterWorkflow(TriggeredSweepWorkflow)
	w.RegisterWorkflow(DailySweepWorkflow)
	w.RegisterWorkflow(RelayExecutorJobsWorkflow)
	w.RegisterWorkflow(ExecuteApprovalWorkflow)
	w.RegisterActivityWithOptions(opts.Activities.ExpireSnoozes,
		activityOptions(ExpireSnoozesActivityName))
	w.RegisterActivityWithOptions(opts.Activities.ListUndelivered,
		activityOptions(ListUndeliveredActivityName))
	w.RegisterActivityWithOptions(opts.Activities.StartDeliveries,
		activityOptions(StartDeliveriesActivityName))
	w.RegisterActivityWithOptions(opts.Activities.DeliverMessage,
		activityOptions(DeliverMessageActivityName))
	w.RegisterActivityWithOptions(opts.Activities.ReclaimMessages,
		activityOptions(ReclaimMessagesActivityName))
	w.RegisterActivityWithOptions(opts.Activities.PlanNotification,
		activityOptions(PlanNotificationActivityName))
	w.RegisterActivityWithOptions(opts.Activities.NotifyRecipients,
		activityOptions(NotifyRecipientsActivityName))
	w.RegisterActivityWithOptions(opts.Activities.SettleNotification,
		activityOptions(SettleNotificationActivityName))
	w.RegisterActivityWithOptions(opts.Activities.ListSweepTriggers,
		activityOptions(ListSweepTriggersActivityName))
	w.RegisterActivityWithOptions(opts.Activities.StartTriggeredSweeps,
		activityOptions(StartTriggeredSweepsActivityName))
	w.RegisterActivityWithOptions(opts.Activities.RunWatcher,
		activityOptions(RunWatcherActivityName))
	w.RegisterActivityWithOptions(opts.Activities.RunAnalyst,
		activityOptions(RunAnalystActivityName))
	w.RegisterActivityWithOptions(opts.Activities.SettleSweepTrigger,
		activityOptions(SettleSweepTriggerActivityName))
	w.RegisterActivityWithOptions(opts.Activities.ListSweepTargets,
		activityOptions(ListSweepTargetsActivityName))
	w.RegisterActivityWithOptions(opts.Activities.NextFindingToNarrate,
		activityOptions(NextFindingToNarrateActivityName))
	w.RegisterActivityWithOptions(opts.Activities.RecordNarrative,
		activityOptions(RecordNarrativeActivityName))
	w.RegisterActivityWithOptions(opts.Activities.ListExecutorJobs,
		activityOptions(ListExecutorJobsActivityName))
	w.RegisterActivityWithOptions(opts.Activities.StartExecutions,
		activityOptions(StartExecutionsActivityName))
	w.RegisterActivityWithOptions(opts.Activities.ExecuteJob,
		activityOptions(ExecuteJobActivityName))
	w.RegisterActivityWithOptions(opts.Activities.LoadWatchContext,
		activityOptions(LoadWatchContextActivityName))
	// DraftNarrative and Watch are deliberately NOT registered here: both run
	// on the `intelligence` task queue, served by the Python worker (§16.4).

	if err := w.Start(); err != nil {
		return nil, fmt.Errorf("schedule: starting the worker: %w", err)
	}

	for _, def := range schedules(opts) {
		if err := ensureSchedule(ctx, c, opts, def); err != nil {
			w.Stop()
			return nil, err
		}
	}

	opts.Logger.Info("temporal worker started",
		"task_queue", opts.TaskQueue,
		"namespace", opts.Namespace,
		"snooze_expiry", opts.SnoozeExpirySchedule,
		"outbox_relay", opts.OutboxRelayInterval.String(),
		"outbox_reclaim", opts.OutboxReclaimSchedule,
		"sweep_relay", opts.SweepRelayInterval.String(),
		"sweep", opts.SweepSchedule,
		"executor_relay", opts.ExecutorRelayInterval.String())
	return &Worker{client: c, worker: w}, nil
}

// Stop stops polling and closes the client.
func (w *Worker) Stop() {
	w.worker.Stop()
	w.client.Close()
}

// Client is the connected client, for the readiness probe.
func (w *Worker) Client() client.Client { return w.client }

// Ready reports whether the engine answers, which is what /readyz asks.
//
// This is the probe ENT-256 step 1 said core-api would carry. It is here
// instead because this is the process that depends on the engine, and a
// readiness check belongs to the process that would be not ready without the
// thing it checks.
func Ready(c client.Client) func(context.Context) error {
	return func(ctx context.Context) error {
		ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
		defer cancel()
		if _, err := c.CheckHealth(ctx, &client.CheckHealthRequest{}); err != nil {
			return fmt.Errorf("temporal: %w", err)
		}
		return nil
	}
}

// scheduleDefinition is one Schedule as this worker wants it to exist: what
// fires when, and what it starts.
type scheduleDefinition struct {
	ID   string
	Spec client.ScheduleSpec
	// Workflow is the function the schedule starts, on this worker's queue.
	Workflow any
	// ExecutionTimeout bounds one run, retries included. Under the interval
	// between ticks, so the overlap policy never fires in practice.
	ExecutionTimeout time.Duration
	// CatchupWindow is how far back a missed tick (the engine was down) is
	// made up. Every workflow here is idempotent, so one catch-up run does
	// the work of every missed one; what this bounds is how stale a tick
	// still gets run at all.
	CatchupWindow time.Duration
	Memo          map[string]any
}

// schedules is every Schedule this worker owns, in one place, so the list the
// CI step asserts against and the list the code registers are the same list.
func schedules(opts Options) []scheduleDefinition {
	return []scheduleDefinition{
		{
			ID: SnoozeExpiryScheduleID,
			// A missed tick (the engine was down) runs once when it is back,
			// rather than once per missed hour. The pass is idempotent, so
			// running it ten times for ten missed hours would be ten UPDATEs
			// that do the work of one; running it once is the same outcome.
			Spec:             client.ScheduleSpec{CronExpressions: []string{opts.SnoozeExpirySchedule}},
			Workflow:         ExpireSnoozedFindingsWorkflow,
			ExecutionTimeout: 50 * time.Minute,
			CatchupWindow:    time.Hour,
			Memo: map[string]any{
				"owner": "workers (ENT-256)",
				"what":  "brings deferred findings back when their date passes",
			},
		},
		{
			ID: OutboxRelayScheduleID,
			// An interval rather than a cron, because a cron's finest grain is
			// a minute and mail a person is waiting for should leave in
			// seconds. The catch-up window is deliberately short: a tick
			// missed while the engine was down is made up by the next tick
			// listing the same rows, so replaying old ticks would only start
			// the same deliveries a few more times and find them running.
			Spec: client.ScheduleSpec{
				Intervals: []client.ScheduleIntervalSpec{{Every: opts.OutboxRelayInterval}},
			},
			Workflow:         RelayOutboxWorkflow,
			ExecutionTimeout: 5 * time.Minute,
			CatchupWindow:    time.Minute,
			Memo: map[string]any{
				"owner": "workers (ENT-256)",
				"what":  "starts a delivery for every message and every finding notification waiting to leave",
			},
		},
		{
			ID:               OutboxReclaimScheduleID,
			Spec:             client.ScheduleSpec{CronExpressions: []string{opts.OutboxReclaimSchedule}},
			Workflow:         ReclaimOutboxWorkflow,
			ExecutionTimeout: 50 * time.Minute,
			CatchupWindow:    time.Hour,
			Memo: map[string]any{
				"owner": "workers (ENT-256)",
				"what":  "clears addresses and bodies out of delivered and abandoned messages",
			},
		},
		{
			ID: SweepTriggerRelayScheduleID,
			// The same interval shape as the outbox relay, for the same
			// reason: somebody just confirmed onboarding and is looking at an
			// empty feed.
			Spec: client.ScheduleSpec{
				Intervals: []client.ScheduleIntervalSpec{{Every: opts.SweepRelayInterval}},
			},
			Workflow:         RelaySweepTriggersWorkflow,
			ExecutionTimeout: 5 * time.Minute,
			CatchupWindow:    time.Minute,
			Memo: map[string]any{
				"owner": "workers (ENT-256)",
				"what":  "starts a sweep for every organisation that asked for one (confirmed onboarding)",
			},
		},
		{
			ID: ExecutorRelayScheduleID,
			// The same interval shape as the outbox and sweep relays, and
			// the tightest of the three reasons: somebody clicked approve
			// and is looking at Records for what they approved.
			Spec: client.ScheduleSpec{
				Intervals: []client.ScheduleIntervalSpec{{Every: opts.ExecutorRelayInterval}},
			},
			Workflow:         RelayExecutorJobsWorkflow,
			ExecutionTimeout: 5 * time.Minute,
			CatchupWindow:    time.Minute,
			Memo: map[string]any{
				"owner": "workers (ENT-271)",
				"what":  "creates the record every approved finding asked for",
			},
		},
		{
			ID:       DailySweepScheduleID,
			Spec:     client.ScheduleSpec{CronExpressions: []string{opts.SweepSchedule}},
			Workflow: DailySweepWorkflow,
			// The estate, four organisations at a time, seconds each: an hour
			// is generous for thousands. Under a day, so the overlap policy
			// never fires.
			ExecutionTimeout: 12 * time.Hour,
			CatchupWindow:    6 * time.Hour,
			Memo: map[string]any{
				"owner": "workers (ENT-256)",
				"what":  "runs the Watcher and then the Analyst over every organisation with a profile",
			},
		},
	}
}

// ensureSchedule creates the Schedule, or brings an existing one into line
// with the configured spec and action, retrying while the engine finishes
// starting.
//
// Create-then-update rather than delete-and-recreate, because a Schedule has
// state worth keeping: its history of runs, whether an operator paused it,
// and the backfill it may be part way through. Changing the cron in
// configuration should change the cron and nothing else.
//
// # WHY IT RETRIES, MEASURED RATHER THAN GUESSED
//
// The engine's healthcheck is `temporal operator cluster health`, which is
// the frontend answering. For a few seconds after a restart the frontend is
// up and the matching and history services have not yet joined its ring, and
// a schedule update in that window fails with "Not enough hosts to serve the
// request" and then the SDK's ten-second deadline. The first version of this
// returned that error, the binary exited 1, and because core-api waits on
// this service being healthy the whole stack failed to come up. It passed on
// a fresh stack, where the create path happened to land after the ring had
// formed, and failed in CI's air-gap job, which is the one place the stack is
// brought up twice in a row on the same volumes and so took the update path
// against an engine seconds old. Same class as ENT-253: a first-minute race
// that a laptop stack is always past.
//
// So: keep trying for two minutes, with the per-attempt deadline the SDK
// already applies, and only then fail, loudly, because a deployment with no
// schedule is the silent failure the CI step exists to catch.
func ensureSchedule(ctx context.Context, c client.Client, opts Options, def scheduleDefinition) error {
	const attempts = 60
	var lastErr error
	for attempt := 1; attempt <= attempts; attempt++ {
		err := ensureScheduleOnce(ctx, c, opts, def)
		if err == nil {
			return nil
		}
		lastErr = err
		opts.Logger.Info("waiting for temporal to accept the schedule",
			"id", def.ID, "attempt", attempt, "error", err)
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(2 * time.Second):
		}
	}
	return fmt.Errorf("schedule: %s could not be registered after %d attempts: %w",
		def.ID, attempts, lastErr)
}

// ensureScheduleOnce is one attempt; see the caller for why there are several.
func ensureScheduleOnce(ctx context.Context, c client.Client, opts Options, def scheduleDefinition) error {
	spec := def.Spec
	spec.Jitter = 0
	action := &client.ScheduleWorkflowAction{
		ID:                       def.ID + "-run",
		Workflow:                 def.Workflow,
		TaskQueue:                opts.TaskQueue,
		WorkflowExecutionTimeout: def.ExecutionTimeout,
	}

	handle := c.ScheduleClient().GetHandle(ctx, def.ID)
	_, err := c.ScheduleClient().Create(ctx, client.ScheduleOptions{
		ID:     def.ID,
		Spec:   spec,
		Action: action,
		// If a run is still going when the next tick comes, skip the tick:
		// every workflow here is idempotent and the running one will do the
		// work.
		Overlap:       enumspb.SCHEDULE_OVERLAP_POLICY_SKIP,
		CatchupWindow: def.CatchupWindow,
		Memo:          def.Memo,
	})
	if err == nil {
		opts.Logger.Info("schedule created", "id", def.ID)
		return nil
	}
	if !errors.Is(err, temporal.ErrScheduleAlreadyRunning) {
		return fmt.Errorf("schedule: creating %s: %w", def.ID, err)
	}

	// It exists. Make it match the configuration, in case the cron changed.
	err = handle.Update(ctx, client.ScheduleUpdateOptions{
		DoUpdate: func(in client.ScheduleUpdateInput) (*client.ScheduleUpdate, error) {
			in.Description.Schedule.Spec = &spec
			in.Description.Schedule.Action = action
			return &client.ScheduleUpdate{Schedule: &in.Description.Schedule}, nil
		},
	})
	if err != nil {
		return fmt.Errorf("schedule: updating %s: %w", def.ID, err)
	}
	opts.Logger.Info("schedule exists, configuration applied", "id", def.ID)
	return nil
}

File: apps/intelligence/src/kindlast_intelligence/worker.py
"""The Temporal worker: this service's half of the `intelligence` task queue
(ENT-256, part five; core-api-surface §16.4).

# WHAT RUNS HERE

Three activities, each the same run the RPC of its name performs: the harness,
the guardrail ring, whatever validator that skill has, and the run record. The
Go workflows load, this service drafts or decides, and Go persists, each half
retrying on its own.

  DraftNarrative   the Analyst. The sweep workflow asks core-api for the next
                   finding with no narrative and records what comes back.
  Watch            the Watcher, over one organisation's context.
  DraftMessage     the Messenger (ENT-260). The doorbell workflow runs it
                   between planning who to tell and telling them, and treats
                   anything other than a drafted message as "use the
                   template", so a notification goes out either way.

# WHAT THIS WORKER HOLDS, WHICH IS THE SAME AS THE RPC: NOTHING

No database handle, no tenancy GUC, no model endpoint, no credential. Every
model call goes back through core-api's CompletionService bound to the
organisation the draft is for, and the run is recorded through core-api's
RecordAgentRun, with this service's own OAuth client, which is the one
credential it legitimately has. The draft request arrives from the engine,
built by core-api; it is data, and this worker cannot be asked to draft for
an organisation core-api did not name.

# WHY THE ACTIVITY IS SYNCHRONOUS

The harness is synchronous and blocks on the model for seconds to minutes;
Temporal's Python SDK runs such activities on a thread pool, sized here to
what one local model can serve at once (ENT-235: one or two requests), so a
queue of drafts is a queue rather than a pile of timeouts. The RPC half of
this process serves from waitress beside it.
"""

from __future__ import annotations

import asyncio
import logging
import threading
from concurrent.futures import ThreadPoolExecutor

from connectrpc.code import Code
from connectrpc.errors import ConnectError
from kindlast.platform.v1 import intelligence_pb2, watcher_pb2
from temporalio import activity
from temporalio.client import Client
from temporalio.exceptions import ApplicationError
from temporalio.worker import Worker

from .service import IntelligenceService

logger = logging.getLogger("kindlast.intelligence.worker")

# Pinned here and in apps/workers/internal/schedule/narrate.go. The Go
# workflow schedules the activity by this name on this queue; the two must
# agree, which the end-to-end run is what proves.
TASK_QUEUE = "intelligence"
ACTIVITY_NAME = "DraftNarrative"
WATCH_ACTIVITY_NAME = "Watch"
# Pinned here first (ENT-260). The Go half of the pin arrives with the caller:
# the doorbell workflow does not schedule this activity yet, and when it does,
# apps/workers/internal/schedule/doorbell.go must name exactly this string,
# which the end-to-end run is what proves.
DRAFT_MESSAGE_ACTIVITY_NAME = "DraftMessage"


def make_activity(service: IntelligenceService):
    """The activity, closed over the service that drafts."""

    @activity.defn(name=ACTIVITY_NAME)
    def draft_narrative(
        request: intelligence_pb2.DraftNarrativeRequest,
    ) -> intelligence_pb2.DraftNarrativeResponse:
        try:
            return service.draft(request)
        except ConnectError as exc:
            if exc.code == Code.INVALID_ARGUMENT:
                # A malformed request is core-api's bug, not a transient
                # condition; retrying it produces the same refusal. Said so,
                # so the workflow records it and moves on.
                raise ApplicationError(str(exc), type="bad-request", non_retryable=True) from exc
            # Anything else (the run could not be recorded, the model could
            # not be reached through core-api) is worth the workflow's retry
            # policy.
            raise ApplicationError(str(exc), type=exc.code.name) from exc

    return draft_narrative


def make_watch_activity(service: IntelligenceService):
    """The Watcher activity, closed over the service that watches."""

    @activity.defn(name=WATCH_ACTIVITY_NAME)
    def watch(request: intelligence_pb2.WatchRequest) -> intelligence_pb2.WatchResponse:
        try:
            return service.run_watch(request)
        except ConnectError as exc:
            if exc.code == Code.INVALID_ARGUMENT:
                # A malformed request, or an organisation with no profile.
                # Neither changes by waiting: retrying produces the same
                # refusal, so the workflow records it and moves on.
                raise ApplicationError(str(exc), type="bad-request", non_retryable=True) from exc
            raise ApplicationError(str(exc), type=exc.code.name) from exc

    return watch


def make_draft_message_activity(service: IntelligenceService):
    """The Messenger activity, closed over the service that drafts (ENT-260).

    The doorbell workflow runs this between planning and sending, and treats
    every failure as "use the template". So the two error shapes matter less
    here than anywhere else on this queue: whatever comes back, a notification
    still goes out. What they buy is a workflow history that says WHY the
    template was used, which is the difference between a deployment whose
    Messenger is off and one whose Messenger is broken.
    """

    @activity.defn(name=DRAFT_MESSAGE_ACTIVITY_NAME)
    def draft_message(
        request: intelligence_pb2.DraftMessageRequest,
    ) -> intelligence_pb2.DraftMessageResponse:
        try:
            return service.run_draft_message(request)
        except ConnectError as exc:
            if exc.code == Code.INVALID_ARGUMENT:
                # A malformed request, or a context with no severity. Neither
                # changes by waiting, and a doorbell must not sit in a retry
                # loop over the words it would have been nicer to use.
                raise ApplicationError(str(exc), type="bad-request", non_retryable=True) from exc
            raise ApplicationError(str(exc), type=exc.code.name) from exc

    return draft_message


async def run_worker(
    service: IntelligenceService,
    address: str,
    namespace: str,
    concurrency: int,
    stop: threading.Event,
) -> None:
    """Connect, retrying while the engine finishes starting, then poll until
    asked to stop."""
    client = None
    attempt = 0
    while client is None and not stop.is_set():
        attempt += 1
        try:
            client = await Client.connect(address, namespace=namespace)
        except Exception as exc:  # noqa: BLE001 (the engine is not up yet)
            logger.info("waiting for temporal at %s (attempt %d): %s", address, attempt, exc)
            await asyncio.sleep(2)
    if client is None:
        return

    worker = Worker(
        client,
        task_queue=TASK_QUEUE,
        activities=[
            make_activity(service),
            make_watch_activity(service),
            make_draft_message_activity(service),
        ],
        activity_executor=ThreadPoolExecutor(max_workers=concurrency),
        max_concurrent_activities=concurrency,
    )
    logger.info(
        "temporal worker started task_queue=%s namespace=%s concurrency=%d",
        TASK_QUEUE,
        namespace,
        concurrency,
    )
    async with worker:
        while not stop.is_set():
            await asyncio.sleep(1)
    logger.info("temporal worker stopped")


def start_in_background(
    service: IntelligenceService, address: str, namespace: str, concurrency: int
) -> threading.Event:
    """Run the worker on its own thread with its own event loop, beside the
    WSGI server. Returns the event that stops it."""
    stop = threading.Event()

    def main() -> None:
        try:
            asyncio.run(run_worker(service, address, namespace, concurrency, stop))
        except Exception:  # noqa: BLE001
            logger.exception("the temporal worker stopped with an error; the RPC half keeps serving")

    thread = threading.Thread(target=main, name="temporal-worker", daemon=True)
    thread.start()
    return stop
